File: src/rogue/interfaces/api/Variable.cpp
/**
 *-----------------------------------------------------------------------------
 * Title      : C++ API Variable
 * ----------------------------------------------------------------------------
 * File       : Variable.cpp
 * Created    : 2023-04-17
 * ----------------------------------------------------------------------------
 * This file is part of the rogue software platform. It is subject to
 * the license terms in the LICENSE.txt file found in the top-level directory
 * of this distribution and at:
 *    https://confluence.slac.stanford.edu/display/ppareg/LICENSE.html.
 * No part of the rogue software platform, including this file, may be
 * copied, modified, propagated, or distributed except according to the terms
 * contained in the LICENSE.txt file.
 * ----------------------------------------------------------------------------
**/
#include <rogue/interfaces/api/Node.h>
#include <rogue/interfaces/api/Variable.h>
#include <boost/make_shared.hpp>
#include <rogue/GeneralError.h>

namespace bp = boost::python;
namespace ria = rogue::interfaces::api;


// Class factory which returns a pointer to a Variable (VariablePtr)
ria::VariablePtr ria::Variable::create (bp::object obj) {
   ria::VariablePtr r = std::make_shared<ria::Variable>(obj);
   return(b);
}

ria::Variable::Variable (boost::python::object obj) {
   _obj = obj;
}

ria::Variable::~Variable() { }


//! Get type string
std::string ria::Variable::typeStr() {
   return(std::string(bp::extract<char *>(this->_obj.attr("typeStr"))));
}

//! Get precision
int32_t ria::Variable::precision() {
   return(bp::extract<uint32_t>(this->_obj.attr("typeStr")));
}

//! Get enum mapping in yaml format, empty string if no enum
std::string ria::Variable::enumYaml() {
   return(std::string(bp::extract<char *>(this->_obj.attr("enumYaml"))));
}

//! Get mode
std::string ria::Variable::mode() {
   return(std::string(bp::extract<char *>(this->_obj.attr("mode"))));
}

//! Get units
std::string ria::Variable::units() {
   return(std::string(bp::extract<char *>(this->_obj.attr("units"))));
}

//! Get minimum
float ria::Variable::minimum() {
   return(std::string(bp::extract<float>(this->_obj.attr("minimum"))));
}

//! Get maximum
float ria::Variable::maximum() {
   return(std::string(bp::extract<float>(this->_obj.attr("maximum"))));
}

//! Set value
void ria::Variable::setDisp(std::string value, bool write=true, int32_t index=-1) {

}

//! Get value
std::string ria::Variable::getDisp(bool read=true, int32_t index=-1) {


}

File: python/pyrogue/interfaces/stream/_Variable.py
#-----------------------------------------------------------------------------
# Company    : SLAC National Accelerator Laboratory
#-----------------------------------------------------------------------------
# Description:
# Class to generate variable update streams
#-----------------------------------------------------------------------------
# This file is part of the rogue software platform. It is subject to
# the license terms in the LICENSE.txt file found in the top-level directory
# of this distribution and at:
#    https://confluence.slac.stanford.edu/display/ppareg/LICENSE.html.
# No part of the rogue software platform, including this file, may be
# copied, modified, propagated, or distributed except according to the terms
# contained in the LICENSE.txt file.
#-----------------------------------------------------------------------------

import pyrogue
import rogue.interfaces.stream
from typing import Any


class Variable(rogue.interfaces.stream.Master):
    """Stream master that emits variable values as YAML frames.

    Parameters
    ----------
    root : pyrogue.Root
        PyRogue root node to monitor.
    incGroups : str | list[str] | None, optional
        Groups to include in variable updates.
    excGroups : str | list[str] | None, optional
        Groups to exclude from variable updates.
    """

    def __init__(
        self,
        *,
        root: pyrogue.Root,
        incGroups: str | list[str] | None = None,
        excGroups: str | list[str] | None = ['NoStream'],
    ) -> None:
        super().__init__()
        self._updateList = {}
        root.addVarListener(func=self._varUpdate, done=self._varDone, incGroups=incGroups, excGroups=excGroups)

        self._root = root
        self._incGroups = incGroups
        self._excGroups = excGroups

    def streamYaml(self) -> None:
        """
        Generate a frame containing all variable values in YAML format.
        A hardware read is not generated before the frame is generated.
        """
        self._sendYamlFrame(self._root.getYaml(readFirst=False,
                                               modes=['RW','RO','WO'],
                                               incGroups=self._incGroups,
                                               excGroups=self._excGroups,
                                               recurse=True))

    def _varUpdate(self, path: str, value: Any) -> None:
        """Queue a variable update for the next batch."""
        self._updateList[path] = value

    def _varDone(self) -> None:
        """Flush queued updates and send as a YAML frame."""
        if len(self._updateList) > 0:
            self._sendYamlFrame(pyrogue.dataToYaml(self._updateList))
            self._updateList = {}

    def _sendYamlFrame(self, yml: str) -> None:
        """Serialize YAML to a frame and send it."""
        b = bytearray(yml,'utf-8')
        frame = self._reqFrame(len(b),True)
        frame.write(b,0)
        self._sendFrame(frame)
